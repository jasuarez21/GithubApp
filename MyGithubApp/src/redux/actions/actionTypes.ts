export default {
    GET_USER: 'GET_USER',
    ERROR: 'ERROR'
}