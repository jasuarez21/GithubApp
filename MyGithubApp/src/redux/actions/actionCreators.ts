import axios from 'axios';
import actionTypes from './actionTypes';
import {Dispatch} from 'redux';

export function getUsers(user: string){
    console.log(user)
    return async (dispatch: Dispatch) => {
        try{
            const {data} = await axios.get('the url')
            dispatch({
                type: actionTypes.GET_USER,
                data,
            })
        } catch (error) {
            dispatch({
                type: actionTypes.ERROR
            })
        }
    }
}