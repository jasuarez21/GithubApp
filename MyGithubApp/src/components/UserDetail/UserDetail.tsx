import React, {useEffect, useState} from 'react';
import { StackScreenProps } from '@react-navigation/stack';
import {useSelector, useDispatch} from 'react-redux';
import { Text, TouchableHighlight, ScrollView, View, StyleSheet, Image} from 'react-native';
import { getFollowers, getRepos } from '../../redux/actions/actionCreators';


interface Props extends StackScreenProps<any, any> {}

const UserDetail = ({route, navigation}: any) => {
    const followers: any = useSelector((store: any) => store.followers);
    const repos: any = useSelector((store: any) => store.repos);
    const {userSelected} = route.params;
    let [init, setInit] = useState(0);
    let [end, setEnd] = useState(5);
    let [page, setPage] = useState(1)
    const dispatch = useDispatch();
    useEffect(() => {
        dispatch(getFollowers(userSelected.followers_url))
        dispatch(getRepos(userSelected.repos_url))
    }, [])
    const nextPage = () => {
        if(repos.slice( init + 5, end + 5).length === 0){
            return false
        } else {
            setPage(page += 1)
            setInit(init += 5);
            setEnd(end += 5)
        }
    }
    const previousPage = () => {
        if(page === 1){
            return false
        } else {
            setPage(page -= 1)
            setInit(init-=5);
            setEnd(end-=5)
        }
    }
    return (
        <ScrollView style={styles.detailContainer}>
        <TouchableHighlight  
            style={styles.backContainer}
            onPress={() => navigation.navigate('ListOfUsers', {})}>
        <Image
                style={styles.iconBack}
                source={{
                    uri: "https://img.icons8.com/color/48/000000/circled-left-2--v1.png",
                }}
            />                          
        </TouchableHighlight>
            <Text style={styles.username}>{userSelected?.login}</Text>
            <Image
                style={styles.photoDetail}
                source={{
                    uri: userSelected?.avatar_url,
                }}
            />
            <Text style={styles.informationUser}>{userSelected?.html_url}</Text>
            <Text style={styles.informationUser}>Tiene {followers.length} seguidores!</Text>
            <View style={styles.reposHeaderContainer}>
                <Text style={styles.titleRepo}>REPOSITORIOS</Text>
                <TouchableHighlight  
                    style={styles.backButton}
                    onPress={() => previousPage()}>
                    <Image
                        style={styles.iconsPaginate}
                        source={{
                            uri: "https://i.postimg.cc/WbLjCbqG/icons8-back-32-2.png",
                        }}
                    />                          
                </TouchableHighlight>
                <Text style={styles.titleRepo}>{page}</Text>
                <TouchableHighlight  
                    style={styles.nextButton}
                    onPress={() => nextPage()}>
                    <Image
                        style={styles.iconsPaginate}
                        source={{
                            uri: "https://i.postimg.cc/wTsFr9Yd/icons8-more-than-50.png",
                        }}
                    />                          
                </TouchableHighlight>
            </View>
            {
                repos.slice(init, end).map((repo: any) => (
                    <View style={styles.repoContainer}>
                        <Text style={styles.nameRepo}>{repo.name}</Text>
                        <Text style={styles.informationRepo}>{repo.description}</Text>
                        <Text style={styles.informationRepo}>{repo.git_url}</Text>
                    </View>
                ))                 
            }
        </ScrollView>
    )}

    const styles = StyleSheet.create({
        detailContainer: {
            display: 'flex',
            flexDirection: 'column'
        },
        backContainer:{
            width: 50,
            position: 'absolute',
            marginTop: 10,
            marginLeft: 10
        },
        iconsPaginate: {
            width: 30,
            height: 30,
            marginBottom: 3 
        },
        backButton: {
            alignSelf: 'flex-end'
        },
        nextButton: {
            alignSelf: 'flex-end'
        },
        reposHeaderContainer: {
            height: 50,
            width: 450,
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'space-between',
            borderBottomWidth: 1,
            borderBottomColor: '#023e8a',
            marginBottom: 10,
            marginLeft: 5,
            marginTop: 5
        },
        iconBack: {
            height: 40,
            width: 40
        },
        repoContainer: {
            backgroundColor: '#caf0f8',
            marginTop: 10,
            width: 350,
            height: 'auto',
            padding: 10,
            alignSelf: 'center',
            borderRadius: 15,
            borderWidth: 1,
            borderColor: '#023e8a'
        },
        titleRepo: {
            fontSize: 18,
            color: '#023e8a',
            marginLeft: 10,
            marginRight: 30,
            marginTop: 20
        },
        nameRepo: {
            fontSize: 16,
            marginLeft: 20,
            textTransform: 'uppercase'
        },
        informationRepo: {
            fontSize: 14,
            marginLeft: 20,
        },
        username: {
            color: '#023e8a',
            fontSize: 28,
            textTransform: 'uppercase',
            alignSelf: 'center',
            marginTop: 20,
            borderBottomWidth: 1,
            borderBottomColor: '#023e8a'
        },
        informationUser: {
            alignSelf: 'center',
            marginTop: 20,
            fontSize: 20
        },
        photoDetail: {
            height: 200,
            width: 200,
            borderRadius: 25,
            marginTop: 20,
            alignSelf: 'center',
            borderWidth: 3,
            borderColor: '#03045e'
        },
        followersContainer: {
            width: 600,
            height: 600,
            borderWidth: 1,
            borderColor: 'red'
        }
    })
    

export default UserDetail