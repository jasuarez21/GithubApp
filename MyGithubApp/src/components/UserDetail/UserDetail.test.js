import { render, fireEvent } from '@testing-library/react-native';
import React from 'react';
import ConfigureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/actionCreators';
import { Provider } from 'react-redux';
import UserDetail from './UserDetail'

jest.mock('../../redux/actions/actionCreators');

describe('When invoked a UserDetail component', () => {
    const mockStore = ConfigureStore();
    const store = mockStore({
        followers: [{
            name: 'Maria'
        }],
        repos: [{
            name: 'GitHubApp'
        }]
    })
    let route;
    let navigation;
    beforeEach(() => {
        jest.spyOn(actions, 'getFollowers').mockReturnValueOnce({ type: '' })
        jest.spyOn(actions, 'getRepos').mockReturnValueOnce({ type: '' })
        navigation = {
            navigate: jest.fn()
        }
        route = {
            params: {
                userSelected : 'Josep'
            }
        }
    })
    test('Should render the UserDetail page', () => {
        const list = render(<Provider store={store}><UserDetail navigation={navigation} route={route} /></Provider>)
        expect(list).toMatchSnapshot()
    })
    test('Should call navigation.navigate when i click backButton button',  () => {
        const { getByTestId } = render(<Provider store={store}><UserDetail navigation={navigation} route={route} /></Provider>);
        const backButton = getByTestId('backButton');
        fireEvent(backButton, 'press', navigation.navigate);
        expect(navigation.navigate).toHaveBeenCalled();
    })
    test('Should not call setRepoEnd when i click previousPage button',  () => {
        const { getByTestId } = render(<Provider store={store}><UserDetail navigation={navigation} route={route} /></Provider>);
        const previousPage = getByTestId('previousRepoPage');
        let setRepoEnd = jest.fn()
        fireEvent(previousPage, 'press', setRepoEnd);
        expect(setRepoEnd).not.toHaveBeenCalled();
    })
    test('Should not call setRepoEnd when i click nextRepoPage button',  () => {
        const { getByTestId } = render(<Provider store={store}><UserDetail navigation={navigation} route={route} /></Provider>);
        const nextRepoPage = getByTestId('nextRepoPage');
        let setRepoEnd = jest.fn()
        fireEvent(nextRepoPage, 'press', setRepoEnd);
        expect(setRepoEnd).not.toHaveBeenCalled();
    })
    test('Should not call setFollowerEnd when i click nextFollowerPage button',  () => {
        const { getByTestId } = render(<Provider store={store}><UserDetail navigation={navigation} route={route} /></Provider>);
        const nextFollowerPage = getByTestId('nextFollowerPage');
        let setFollowerEnd = jest.fn()
        fireEvent(nextFollowerPage, 'press', setFollowerEnd);
        expect(setFollowerEnd).not.toHaveBeenCalled();
    })
    test('Should not call setRepoEnd when i click previousFollowerPage button',  () => {
        const { getByTestId } = render(<Provider store={store}><UserDetail navigation={navigation} route={route} /></Provider>);
        const previousFollowerPage = getByTestId('previousFollowerPage');
        let setFollowerEnd = jest.fn()
        fireEvent(previousFollowerPage, 'press', setFollowerEnd);
        expect(setFollowerEnd).not.toHaveBeenCalled();
    })
})