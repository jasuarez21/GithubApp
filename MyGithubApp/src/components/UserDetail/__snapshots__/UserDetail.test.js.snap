// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When invoked a UserDetail component Should render the UserDetail page 1`] = `
<RCTScrollView
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": 10,
          "marginTop": 10,
          "position": "absolute",
          "width": 50,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "https://img.icons8.com/color/48/000000/circled-left-2--v1.png",
          }
        }
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
        testID="backButton"
      />
    </View>
    <Text
      style={
        Object {
          "alignSelf": "center",
          "borderBottomColor": "#023e8a",
          "borderBottomWidth": 1,
          "color": "#023e8a",
          "fontSize": 28,
          "marginTop": 20,
          "textTransform": "uppercase",
        }
      }
    />
    <Image
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "borderColor": "#03045e",
          "borderRadius": 25,
          "borderWidth": 3,
          "height": 200,
          "marginTop": 20,
          "width": 200,
        }
      }
    />
    <Text
      style={
        Object {
          "alignSelf": "center",
          "fontSize": 20,
          "marginTop": 20,
        }
      }
    />
    <View
      style={
        Object {
          "borderBottomColor": "#023e8a",
          "borderBottomWidth": 1,
          "display": "flex",
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "space-between",
          "marginBottom": 10,
          "marginLeft": 5,
          "marginTop": 5,
          "width": 450,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#023e8a",
            "fontSize": 18,
            "marginLeft": 10,
            "marginRight": 30,
            "marginTop": 20,
          }
        }
      >
        REPOSITORIOS
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "https://i.postimg.cc/WbLjCbqG/icons8-back-32-2.png",
            }
          }
          style={
            Object {
              "height": 30,
              "marginBottom": 3,
              "width": 30,
            }
          }
          testID="previousRepoPage"
        />
      </View>
      <Text
        style={
          Object {
            "color": "#023e8a",
            "fontSize": 18,
            "marginLeft": 10,
            "marginRight": 30,
            "marginTop": 20,
          }
        }
      >
        1
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "flex-end",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "https://i.postimg.cc/wTsFr9Yd/icons8-more-than-50.png",
            }
          }
          style={
            Object {
              "height": 30,
              "marginBottom": 3,
              "width": 30,
            }
          }
          testID="nextRepoPage"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "#caf0f8",
          "borderColor": "#023e8a",
          "borderRadius": 15,
          "borderWidth": 1,
          "height": "auto",
          "marginTop": 10,
          "padding": 10,
          "width": 350,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 16,
            "marginLeft": 20,
            "textTransform": "uppercase",
          }
        }
      >
        GitHubApp
      </Text>
      <Text
        style={
          Object {
            "fontSize": 14,
            "marginLeft": 20,
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 14,
            "marginLeft": 20,
          }
        }
      />
    </View>
    <View>
      <View
        style={
          Object {
            "borderBottomColor": "#023e8a",
            "borderBottomWidth": 1,
            "display": "flex",
            "flexDirection": "row",
            "height": "auto",
            "justifyContent": "space-between",
            "marginBottom": 10,
            "marginLeft": 5,
            "marginTop": 5,
            "padding": 5,
            "width": 450,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#023e8a",
              "fontSize": 18,
              "marginLeft": 10,
              "marginRight": 30,
              "marginTop": 20,
            }
          }
        >
          SEGUIDORES
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignSelf": "flex-end",
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://i.postimg.cc/WbLjCbqG/icons8-back-32-2.png",
              }
            }
            style={
              Object {
                "height": 30,
                "marginBottom": 3,
                "width": 30,
              }
            }
            testID="previousFollowerPage"
          />
        </View>
        <Text
          style={
            Object {
              "color": "#023e8a",
              "fontSize": 18,
              "marginLeft": 10,
              "marginRight": 30,
              "marginTop": 20,
            }
          }
        >
          1
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignSelf": "flex-end",
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://i.postimg.cc/wTsFr9Yd/icons8-more-than-50.png",
              }
            }
            style={
              Object {
                "height": 30,
                "marginBottom": 3,
                "width": 30,
              }
            }
            testID="nextFollowerPage"
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#caf0f8",
            "borderRadius": 10,
            "borderWidth": 1,
            "fontSize": 20,
            "height": 60,
            "margin": 5,
            "padding": 10,
            "textAlign": "center",
            "textTransform": "uppercase",
            "width": 180,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
